File: backend/src/svap/stages/stage0a_discovery.py
"""
Stage 0A: Case Discovery — Feed Monitoring

Checks configured enforcement source listing pages for new entries,
fetches each new entry, evaluates its structural richness using LLM,
and queues rich documents for ingestion into the pipeline.

Sources scoring >= 0.7 richness are auto-accepted and ingested.
Sources scoring 0.4-0.7 are queued for human review.
Sources below 0.4 are auto-rejected.

This is a standalone operation — not chained into stages 0-6. It writes to
source_candidates and enforcement_sources. Stage 0 picks up new sources on
the next pipeline run.

Input:  source_feeds table (listing page configurations)
Output: source_candidates (discovered + scored entries),
        enforcement_sources + documents/chunks (for accepted entries)
"""

import hashlib
import re
import time
from datetime import UTC, datetime
from html.parser import HTMLParser
from urllib.parse import urljoin

from svap.bedrock_client import BedrockClient
from svap.rag import DocumentIngester
from svap.stages.stage0_source_fetch import _extract_text, _fetch_url, _is_binary_content
from svap.storage import SVAPStorage

LINK_EXTRACTION_SYSTEM = (
    "You are an analyst identifying enforcement case document links "
    "on government agency listing pages. Be precise — only identify links "
    "that lead to individual case documents, not to other listing pages "
    "or general information."
)

RICHNESS_SYSTEM = (
    "You are an analyst evaluating enforcement documents for a healthcare fraud "
    "analysis pipeline. You assess whether a document contains sufficient structural "
    "detail about fraud schemes to be useful for case extraction."
)

# Stage number for logging (standalone, before stage 0)


class _HTMLLinkExtractor(HTMLParser):
    """Extract all links (<a href>) from HTML."""

    def __init__(self):
        super().__init__()
        self.links: list[dict] = []
        self._current_href: str | None = None
        self._current_text: list[str] = []

    def handle_starttag(self, tag, attrs):
        if tag == "a":
            href = dict(attrs).get("href")
            if href:
                self._current_href = href
                self._current_text = []

    def handle_data(self, data):
        if self._current_href is not None:
            self._current_text.append(data.strip())

    def handle_endtag(self, tag):
        if tag == "a" and self._current_href:
            self.links.append({
                "url": self._current_href,
                "text": " ".join(self._current_text).strip(),
            })
            self._current_href = None


def _extract_links(html: str) -> list[dict]:
    """Extract all links from HTML."""
    parser = _HTMLLinkExtractor()
    parser.feed(html)
    return parser.links


def _filter_links_by_selector(links: list[dict], base_url: str, link_selector: str | None) -> list[dict]:
    """Filter links using a regex selector pattern, resolving relative URLs."""
    result = []
    for link in links:
        url = urljoin(base_url, link["url"])
        link["url"] = url
        if link_selector and re.search(link_selector, url):
            result.append(link)
    return result


def _extract_links_via_llm(
    client: BedrockClient,
    links: list[dict],
    page_text: str,
    feed: dict,
) -> list[dict]:
    """Use LLM to identify enforcement document links from a listing page."""
    link_list = "\n".join(
        f"- [{link['text'][:80]}]({link['url']})" for link in links[:100]
    )
    prompt = client.render_prompt(
        "discovery_extract_links.txt",
        feed_name=feed["name"],
        content_type=feed["content_type"],
        base_url=feed["listing_url"],
        link_list=link_list,
        page_text=page_text[:3000],
    )
    result = client.invoke_json(prompt, system=LINK_EXTRACTION_SYSTEM, temperature=0.1, max_tokens=2000)
    if isinstance(result, list):
        return result
    return result.get("links", [])


def _check_feed(
    feed: dict, storage: SVAPStorage, client: BedrockClient, config: dict
) -> list[dict]:
    """Check a single feed listing page for new candidate URLs."""
    listing_url = feed["listing_url"]
    max_per_feed = config.get("discovery", {}).get("max_candidates_per_feed", 50)

    print(f"  Checking feed: {feed['name']}")
    try:
        html = _fetch_url(listing_url)
    except Exception as e:
        print(f"    Failed to fetch listing page: {e}")
        return []

    raw_links = _extract_links(html)
    page_text = _extract_text(html)

    # Try regex selector first for a fast path
    filtered = _filter_links_by_selector(raw_links, listing_url, feed.get("link_selector"))

    # If selector matched nothing or no selector, use LLM
    if not filtered:
        resolved = [{"url": urljoin(listing_url, lnk["url"]), "text": lnk["text"]} for lnk in raw_links]
        filtered = _extract_links_via_llm(client, resolved, page_text, feed)

    # Dedup against existing candidates
    new_candidates = []
    for entry in filtered[:max_per_feed]:
        url = entry.get("url", "")
        if not url:
            continue
        existing = storage.get_candidate_by_url(url)
        if existing:
            continue

        candidate_id = hashlib.sha256(url.encode()).hexdigest()[:12]
        candidate = {
            "candidate_id": candidate_id,
            "feed_id": feed["feed_id"],
            "title": entry.get("title") or entry.get("text", "Untitled")[:200],
            "url": url,
            "discovered_at": datetime.now(UTC).isoformat(),
            "published_date": entry.get("published_date"),
            "status": "discovered",
        }
        storage.insert_candidate(candidate)
        new_candidates.append(candidate)

    print(f"    Found {len(filtered)} links, {len(new_candidates)} new candidates")
    return new_candidates


def _evaluate_richness(
    client: BedrockClient, candidate: dict, text: str
) -> dict:
    """Score a document's structural richness via LLM."""
    prompt = client.render_prompt(
        "discovery_richness.txt",
        title=candidate["title"],
        url=candidate["url"],
        text_sample=text[:6000],
    )
    result = client.invoke_json(prompt, system=RICHNESS_SYSTEM, temperature=0.1, max_tokens=500)
    return {
        "richness_score": float(result.get("richness_score", 0.0)),
        "rationale": result.get("rationale", ""),
        "estimated_cases": int(result.get("estimated_cases", 1)),
        "scheme_types": result.get("scheme_types", []),
    }


def _apply_disposition(eval_result: dict, config: dict) -> str:
    """Determine accept/reject/scored based on richness thresholds."""
    score = eval_result["richness_score"]
    accept = config.get("discovery", {}).get("richness_accept_threshold", 0.7)
    review = config.get("discovery", {}).get("richness_review_threshold", 0.4)
    if score >= accept:
        return "accepted"
    if score >= review:
        return "scored"  # queued for human review
    return "rejected"


def _ingest_candidate(
    candidate: dict, text: str, storage: SVAPStorage, config: dict
) -> dict:
    """Ingest an accepted candidate into enforcement_sources and RAG.

    Checks for an existing enforcement source with the same URL to avoid
    duplicating documents when the same URL enters via multiple paths.
    """
    # Check if an enforcement source with this URL already exists
    existing = storage.get_enforcement_source_by_url(candidate["url"])
    if existing and existing.get("has_document") and existing.get("doc_id"):
        # Reuse existing source and document — no re-ingestion needed
        source_id = existing["source_id"]
        doc_id = existing["doc_id"]
        storage.update_candidate_ingested(candidate["candidate_id"], source_id, doc_id)
        return {"doc_id": doc_id, "n_chunks": 0, "source_id": source_id}

    source_id = existing["source_id"] if existing else f"disc_{candidate['candidate_id']}"
    ingester = DocumentIngester(storage, config)

    doc_id, n_chunks = ingester.ingest_text(
        text=text,
        filename=source_id,
        doc_type="enforcement",
        metadata={
            "url": candidate["url"],
            "source_name": candidate["title"],
            "candidate_id": candidate["candidate_id"],
            "discovery_date": datetime.now(UTC).isoformat(),
        },
    )

    storage.upsert_enforcement_source({
        "source_id": source_id,
        "name": candidate["title"],
        "url": candidate["url"],
        "source_type": "discovery",
        "description": f"Discovered from feed. Richness: {candidate.get('richness_score', 'N/A')}",
        "has_document": True,
        "doc_id": doc_id,
        "validation_status": "pending",
        "candidate_id": candidate["candidate_id"],
        "feed_id": candidate.get("feed_id"),
    })

    storage.update_candidate_ingested(candidate["candidate_id"], source_id, doc_id)
    return {"doc_id": doc_id, "n_chunks": n_chunks, "source_id": source_id}


def run(storage: SVAPStorage, client: BedrockClient, config: dict) -> dict:
    """Execute case discovery: check feeds for new enforcement documents.

    Discovery operates independently of pipeline runs — it writes to global
    tables (source_feeds, source_candidates, enforcement_sources, documents)
    that any subsequent pipeline run will pick up.

    Returns a summary dict of what was discovered/accepted/rejected.
    """
    print("Case Discovery — Feed Monitoring")

    feeds = storage.get_source_feeds(enabled_only=True)

    if not feeds:
        print("  No source feeds configured.")
        return {"feeds_checked": 0}

    total_discovered = 0
    total_accepted = 0
    total_rejected = 0
    total_review = 0

    for feed in feeds:
        new_candidates = _check_feed(feed, storage, client, config)
        total_discovered += len(new_candidates)

        for candidate in new_candidates:
            # Fetch full document
            try:
                html = _fetch_url(candidate["url"])
                text = _extract_text(html)
            except Exception as e:
                print(f"    Failed to fetch {candidate['url']}: {e}")
                storage.update_candidate_status(candidate["candidate_id"], "error")
                continue

            if len(text) < 200:
                print(f"    Text too short ({len(text)} chars), skipping")
                storage.update_candidate_status(candidate["candidate_id"], "error")
                continue

            if _is_binary_content(text):
                print("    Binary/non-readable content (PDF or image), skipping")
                storage.update_candidate_status(candidate["candidate_id"], "error")
                continue

            storage.update_candidate_status(candidate["candidate_id"], "fetched")

            # Evaluate richness
            eval_result = _evaluate_richness(client, candidate, text)
            storage.update_candidate_richness(
                candidate["candidate_id"],
                eval_result["richness_score"],
                eval_result["rationale"],
                eval_result["estimated_cases"],
            )
            candidate["richness_score"] = eval_result["richness_score"]

            # Apply disposition
            disposition = _apply_disposition(eval_result, config)
            print(
                f"    {candidate['title'][:60]}: "
                f"richness={eval_result['richness_score']:.2f} → {disposition}"
            )

            if disposition == "accepted":
                _ingest_candidate(candidate, text, storage, config)
                total_accepted += 1
            elif disposition == "rejected":
                storage.update_candidate_status(candidate["candidate_id"], "rejected")
                total_rejected += 1
            else:
                total_review += 1

            # Be polite to government servers
            time.sleep(1)

        storage.update_feed_last_checked(feed["feed_id"])

    summary = {
        "feeds_checked": len(feeds),
        "candidates_discovered": total_discovered,
        "candidates_accepted": total_accepted,
        "candidates_rejected": total_rejected,
        "candidates_pending_review": total_review,
    }
    print(
        f"  Discovery complete: {total_discovered} discovered, "
        f"{total_accepted} accepted, {total_rejected} rejected, "
        f"{total_review} pending review."
    )
    return summary

File: backend/shared/src/stages/stage0a_discovery.rs
//! Stage 0A: Case Discovery -- Feed Monitoring
//!
//! Checks configured enforcement source listing pages for new entries,
//! fetches each, evaluates richness, and queues for ingestion.

use chrono::Utc;
use regex::Regex;
use serde_json::json;
use sha2::{Digest, Sha256};
use tokio_postgres::Client;
use tracing::{error, info};

use crate::bedrock::BedrockClient;
use crate::db;
use crate::rag::DocumentIngester;
use crate::stages::stage0_source_fetch::{extract_text, fetch_url, is_binary_content};
use crate::types::{Config, SourceCandidate, SourceFeed};

const RICHNESS_SYSTEM: &str =
    "You are an analyst evaluating enforcement documents for a healthcare fraud \
     analysis pipeline. You assess whether a document contains sufficient structural \
     detail about fraud schemes to be useful for case extraction.";

/// Prompt templates are loaded from the prompts directory.
const DISCOVERY_EXTRACT_LINKS_PROMPT: &str =
    include_str!("../../prompts/discovery_extract_links.txt");
const DISCOVERY_RICHNESS_PROMPT: &str = include_str!("../../prompts/discovery_richness.txt");

type StageResult<T> = Result<T, Box<dyn std::error::Error + Send + Sync>>;

#[derive(Default)]
struct DiscoveryStats {
    discovered: usize,
    accepted: usize,
    rejected: usize,
    review: usize,
}

impl DiscoveryStats {
    fn add(&mut self, other: DiscoveryStats) {
        self.discovered += other.discovered;
        self.accepted += other.accepted;
        self.rejected += other.rejected;
        self.review += other.review;
    }
}

struct DiscoveryThresholds {
    max_per_feed: usize,
    accept: f64,
    review: f64,
}

impl DiscoveryThresholds {
    fn from_config(config: &Config) -> Self {
        let discovery = config.discovery.as_ref();
        Self {
            max_per_feed: discovery
                .and_then(|d| d.max_candidates_per_feed)
                .unwrap_or(50),
            accept: discovery
                .and_then(|d| d.richness_accept_threshold)
                .unwrap_or(0.7),
            review: discovery
                .and_then(|d| d.richness_review_threshold)
                .unwrap_or(0.4),
        }
    }
}

enum CandidateDisposition {
    Accepted,
    Review,
    Rejected,
}

pub async fn run(
    db_client: &Client,
    bedrock: &BedrockClient,
    _run_id: &str,
    config: &Config,
) -> StageResult<serde_json::Value> {
    info!("Case Discovery -- Feed Monitoring");

    let feeds = db::get_source_feeds(db_client, true).await?;
    if feeds.is_empty() {
        info!("No source feeds configured.");
        return Ok(json!({"feeds_checked": 0}));
    }

    let thresholds = DiscoveryThresholds::from_config(config);
    let ingester = DocumentIngester::new(config);
    let link_re = Regex::new(r#"<a\s+[^>]*href\s*=\s*"([^"]*)"[^>]*>([\s\S]*?)</a>"#).unwrap();
    let mut totals = DiscoveryStats::default();

    for feed in &feeds {
        let stats =
            process_feed(db_client, bedrock, &ingester, &thresholds, &link_re, feed).await?;
        totals.add(stats);
    }

    let summary = json!({
        "feeds_checked": feeds.len(),
        "candidates_discovered": totals.discovered,
        "candidates_accepted": totals.accepted,
        "candidates_rejected": totals.rejected,
        "candidates_pending_review": totals.review,
    });
    info!(
        "Discovery complete: {} discovered, {} accepted, {} rejected, {} pending review.",
        totals.discovered, totals.accepted, totals.rejected, totals.review
    );
    Ok(summary)
}

async fn process_feed(
    db_client: &Client,
    bedrock: &BedrockClient,
    ingester: &DocumentIngester,
    thresholds: &DiscoveryThresholds,
    link_re: &Regex,
    feed: &SourceFeed,
) -> StageResult<DiscoveryStats> {
    info!("Checking feed: {}", feed.name);
    let Some((html, page_text)) = fetch_feed_page(feed).await else {
        return Ok(DiscoveryStats::default());
    };

    let raw_links = extract_raw_links(link_re, &html);
    let urls = discover_candidate_urls(bedrock, feed, &raw_links, &page_text).await?;
    let candidates = insert_new_candidates(db_client, feed, &urls, thresholds.max_per_feed).await?;
    info!(
        "Found {} links, {} new candidates",
        urls.len(),
        candidates.len()
    );

    let mut stats =
        evaluate_candidates(db_client, bedrock, ingester, thresholds, &candidates).await?;
    stats.discovered = candidates.len();
    db::update_feed_last_checked(db_client, &feed.feed_id).await?;
    Ok(stats)
}

async fn fetch_feed_page(feed: &SourceFeed) -> Option<(String, String)> {
    match fetch_url(&feed.listing_url).await {
        Ok(html) => {
            let page_text = extract_text(&html);
            Some((html, page_text))
        }
        Err(e) => {
            error!("Failed to fetch listing page: {}", e);
            None
        }
    }
}

fn extract_raw_links(link_re: &Regex, html: &str) -> Vec<(String, String)> {
    link_re
        .captures_iter(html)
        .map(|cap| (cap[1].to_string(), cap[2].to_string()))
        .collect()
}

async fn discover_candidate_urls(
    bedrock: &BedrockClient,
    feed: &SourceFeed,
    raw_links: &[(String, String)],
    page_text: &str,
) -> StageResult<Vec<String>> {
    let urls = selector_urls(feed, raw_links);
    if urls.is_empty() {
        llm_selected_urls(bedrock, feed, raw_links, page_text).await
    } else {
        Ok(urls)
    }
}

fn selector_urls(feed: &SourceFeed, raw_links: &[(String, String)]) -> Vec<String> {
    let Some(selector) = &feed.link_selector else {
        return Vec::new();
    };
    let Ok(sel_re) = Regex::new(selector) else {
        return Vec::new();
    };
    raw_links
        .iter()
        .map(|(url, _)| resolve_url(&feed.listing_url, url))
        .filter(|url| sel_re.is_match(url))
        .collect()
}

async fn llm_selected_urls(
    bedrock: &BedrockClient,
    feed: &SourceFeed,
    raw_links: &[(String, String)],
    page_text: &str,
) -> StageResult<Vec<String>> {
    let link_list = format_link_list(feed, raw_links);
    let prompt = BedrockClient::render_prompt(
        DISCOVERY_EXTRACT_LINKS_PROMPT,
        &[
            ("feed_name", &feed.name),
            ("content_type", &feed.content_type),
            ("base_url", &feed.listing_url),
            ("link_list", &link_list),
            ("page_text", &page_text[..page_text.len().min(3000)]),
        ],
    );
    let result = bedrock
        .invoke_json(
            &prompt,
            "You are an analyst identifying enforcement case document links.",
            Some(0.1),
            Some(2000),
        )
        .await?;
    Ok(extract_llm_urls(&result))
}

fn format_link_list(feed: &SourceFeed, raw_links: &[(String, String)]) -> String {
    raw_links
        .iter()
        .take(100)
        .map(|(url, text)| {
            let resolved = resolve_url(&feed.listing_url, url);
            let clean_text: String = extract_text(text).chars().take(80).collect();
            format!("- [{}]({})", clean_text, resolved)
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn extract_llm_urls(result: &serde_json::Value) -> Vec<String> {
    let links = result
        .as_array()
        .or_else(|| result.get("links").and_then(|links| links.as_array()));
    links
        .into_iter()
        .flatten()
        .filter_map(|link| link.get("url").and_then(|url| url.as_str()))
        .map(String::from)
        .collect()
}

async fn insert_new_candidates(
    db_client: &Client,
    feed: &SourceFeed,
    urls: &[String],
    max_per_feed: usize,
) -> StageResult<Vec<SourceCandidate>> {
    let mut candidates = Vec::new();
    for url in urls.iter().take(max_per_feed).filter(|url| !url.is_empty()) {
        if db::get_candidate_by_url(db_client, url).await?.is_some() {
            continue;
        }
        let candidate = build_candidate(feed, url);
        db::insert_candidate(db_client, &candidate).await?;
        candidates.push(candidate);
    }
    Ok(candidates)
}

fn build_candidate(feed: &SourceFeed, url: &str) -> SourceCandidate {
    let mut hasher = Sha256::new();
    hasher.update(url.as_bytes());
    let candidate_id = format!("{:x}", hasher.finalize())[..12].to_string();
    let now = Utc::now().to_rfc3339();

    SourceCandidate {
        candidate_id,
        feed_id: Some(feed.feed_id.clone()),
        title: "Untitled".to_string(),
        url: url.to_string(),
        discovered_at: now.clone(),
        published_date: None,
        status: "discovered".to_string(),
        richness_score: None,
        richness_rationale: None,
        estimated_cases: None,
        source_id: None,
        doc_id: None,
        reviewed_by: Some("auto".to_string()),
        created_at: now.clone(),
        updated_at: now,
    }
}

async fn evaluate_candidates(
    db_client: &Client,
    bedrock: &BedrockClient,
    ingester: &DocumentIngester,
    thresholds: &DiscoveryThresholds,
    candidates: &[SourceCandidate],
) -> StageResult<DiscoveryStats> {
    let mut stats = DiscoveryStats::default();
    for candidate in candidates {
        let Some(text) = fetch_candidate_text(db_client, candidate).await? else {
            continue;
        };
        let richness = evaluate_richness(db_client, bedrock, candidate, &text).await?;
        match apply_disposition(db_client, ingester, candidate, &text, richness, thresholds).await?
        {
            CandidateDisposition::Accepted => stats.accepted += 1,
            CandidateDisposition::Review => stats.review += 1,
            CandidateDisposition::Rejected => stats.rejected += 1,
        }
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
    }
    Ok(stats)
}

async fn fetch_candidate_text(
    db_client: &Client,
    candidate: &SourceCandidate,
) -> StageResult<Option<String>> {
    let text = match fetch_url(&candidate.url).await {
        Ok(html) => extract_text(&html),
        Err(e) => {
            error!("Failed to fetch {}: {}", candidate.url, e);
            db::update_candidate_status(db_client, &candidate.candidate_id, "error").await?;
            return Ok(None);
        }
    };

    if text.len() < 200 || is_binary_content(&text) {
        db::update_candidate_status(db_client, &candidate.candidate_id, "error").await?;
        return Ok(None);
    }

    db::update_candidate_status(db_client, &candidate.candidate_id, "fetched").await?;
    Ok(Some(text))
}

async fn evaluate_richness(
    db_client: &Client,
    bedrock: &BedrockClient,
    candidate: &SourceCandidate,
    text: &str,
) -> StageResult<f64> {
    let prompt = BedrockClient::render_prompt(
        DISCOVERY_RICHNESS_PROMPT,
        &[
            ("title", &candidate.title),
            ("url", &candidate.url),
            ("text_sample", &text[..text.len().min(6000)]),
        ],
    );
    let eval = bedrock
        .invoke_json(&prompt, RICHNESS_SYSTEM, Some(0.1), Some(500))
        .await?;
    let richness_score = eval
        .get("richness_score")
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);
    let rationale = eval
        .get("rationale")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let estimated_cases = eval
        .get("estimated_cases")
        .and_then(|v| v.as_i64())
        .unwrap_or(1) as i32;

    db::update_candidate_richness(
        db_client,
        &candidate.candidate_id,
        richness_score,
        &rationale,
        estimated_cases,
    )
    .await?;
    Ok(richness_score)
}

async fn apply_disposition(
    db_client: &Client,
    ingester: &DocumentIngester,
    candidate: &SourceCandidate,
    text: &str,
    richness_score: f64,
    thresholds: &DiscoveryThresholds,
) -> StageResult<CandidateDisposition> {
    if richness_score >= thresholds.accept {
        ingest_candidate(db_client, ingester, candidate, text).await?;
        return Ok(CandidateDisposition::Accepted);
    }
    if richness_score >= thresholds.review {
        return Ok(CandidateDisposition::Review);
    }
    db::update_candidate_status(db_client, &candidate.candidate_id, "rejected").await?;
    Ok(CandidateDisposition::Rejected)
}

async fn ingest_candidate(
    db_client: &Client,
    ingester: &DocumentIngester,
    candidate: &SourceCandidate,
    text: &str,
) -> StageResult<()> {
    let metadata = json!({
        "url": candidate.url,
        "source_name": candidate.title,
        "candidate_id": candidate.candidate_id,
    });
    let source_id = format!("disc_{}", candidate.candidate_id);
    let (doc_id, _) = ingester
        .ingest_text(db_client, text, &source_id, "enforcement", Some(&metadata))
        .await?;
    db::update_candidate_ingested(db_client, &candidate.candidate_id, &source_id, &doc_id).await?;
    Ok(())
}

fn resolve_url(base: &str, href: &str) -> String {
    if href.starts_with("http://") || href.starts_with("https://") {
        return href.to_string();
    }
    // Simple URL resolution
    if href.starts_with('/') {
        if let Some(origin) = base
            .find("//")
            .and_then(|i| base[i + 2..].find('/').map(|j| &base[..i + 2 + j]))
        {
            return format!("{}{}", origin, href);
        }
    }
    format!(
        "{}/{}",
        base.trim_end_matches('/'),
        href.trim_start_matches('/')
    )
}
